### PR TITLE
Fix detection of AVX and NEON

### PR DESCRIPTION
**Pull request by carlesfernandez**
After merging of PR #319, xgetbv and NEON are not properly detected in my environments. This PR seems to solve the issue.

Changes:
 * Let the compiler know where to find `volk/volk_common.h` when testing for xgetbv and NEON
 * Make use of `__VOLK_VOLATILE`
 * Replace `<inttypes.h>` by `<stdint.h>` in `volk/volk_common.h`, fixes configuration tests in macOS/clang

### CODE REVIEW

File: include/volk/volk_common.h [hunk 1]
```diff
@@ -87,7 +87,7 @@
 ////////////////////////////////////////////////////////////////////////
 // The bit128 union used by some
 ////////////////////////////////////////////////////////////////////////
-#include <inttypes.h>
+#include <stdint.h>
```
Comment (jdemel) [inline include/volk/volk_common.h:90]: Is this change necessary? Since your PR tries to fix unknown side-effects from a previous commit, I'd like to avoid any new side-effects.

File: lib/CMakeLists.txt [hunk 2]
```diff
@@ -144,12 +144,13 @@ set(HAVE_AVX_CVTPI32_PS 0)
...
+        file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/test_xgetbv.c "#include <volk/volk_common.h>\n unsigned long long _xgetbv(unsigned int index) { unsigned int eax, edx; __VOLK_ASM __VOLK_VOLATILE(\"xgetbv\" : \"=a\"(eax), \"=d\"(edx) : \"c\"(index)); return ((unsigned long long)edx << 32) | eax; } int main (void) { (void) _xgetbv(0); return (0); }")
+        set(_AUX_INCLUDE_FLAG -I${PROJECT_SOURCE_DIR}/include)
     else (NOT MSVC)
         #MSVC defines an intrinsic
         file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/test_xgetbv.c "#include <stdio.h> \n #include <intrin.h> \n int main() { int avxSupported = 0; \n#if (_MSC_FULL_VER >= 160040219) \nint cpuInfo[4]; __cpuid(cpuInfo, 1);\nif ((cpuInfo[2] & (1 << 27) || 0) && (cpuInfo[2] & (1 << 28) || 0)) \n{\nunsigned long long xcrFeatureMask = _xgetbv(_XCR_XFEATURE_ENABLED_MASK);\n   avxSupported = (xcrFeatureMask & 0x6) == 6;}\n#endif \n return 1- avxSupported; }")
     endif(NOT MSVC)
-    execute_process(COMMAND ${CMAKE_C_COMPILER} -o
+    execute_process(COMMAND ${CMAKE_C_COMPILER} ${_AUX_INCLUDE_FLAG} -o
```
Comment (jdemel) [inline lib/CMakeLists.txt:153]: This variable is only defined if `NOT MSVC` but does MSVC work without this flag?

File: lib/CMakeLists.txt [hunk 3]
```diff
@@ -255,9 +256,10 @@ check_c_source_compiles("#include <arm_neon.h>\nint main(){ uint8_t *dest; uint8
 	                neon_compile_result)
 
 if(neon_compile_result)
-    check_c_source_compiles("#include <volk/volk_common.h>\n int main(){__VOLK_ASM volatile(\"vrev32.8 q0, q0\");}"
+    set(CMAKE_REQUIRED_INCLUDES ${PROJECT_SOURCE_DIR}/include)
+    check_c_source_compiles("#include <volk/volk_common.h>\n int main(){__VOLK_ASM __VOLK_VOLATILE(\"vrev32.8 q0, q0\");}"
                             have_neonv7_result )
-    check_c_source_compiles("#include <volk/volk_common.h>\n int main(){__VOLK_ASM volatile(\"sub v1.4s,v1.4s,v1.4s\");}"
+    check_c_source_compiles("#include <volk/volk_common.h>\n int main(){__VOLK_ASM __VOLK_VOLATILE(\"sub v1.4s,v1.4s,v1.4s\");}"
```
Comment (jdemel) [inline lib/CMakeLists.txt:262]: Did we forgot about this `__VOLK_VOLATILE` use in the previous PR?

**Review by jdemel — COMMENTED** [review]
I left some comments for clarification. Besides this PR looks good to me. Also, thanks for your quick fix.

File: include/volk/volk_common.h (hunk 1 above)
Comment (carlesfernandez) [reply include/volk/volk_common.h:90]: Without it, it does not work in macOS. When commenting out https://github.com/carlesfernandez/volk/blob/67f2f6600f7c19867c9e8b48bf31ea662c225b1f/lib/CMakeLists.txt#L156 , without that change, the error is:


```
...
-- x86* CPU detected
In file included from /Users/carlesfernandez/Documents/volk/build/lib/test_xgetbv.c:1:
In file included from /Users/carlesfernandez/Documents/volk/include/volk/volk_common.h:90:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.0/include/inttypes.h:30:15: fatal error: 'inttypes.h' file not found
#include_next <inttypes.h>
              ^~~~~~~~~~~~
1 error generated.
-- Compiler or linker missing xgetbv instruction, Overruled arch avx
...
```

The difference bewteen these includes is explained here https://stackoverflow.com/questions/7597025/difference-between-stdint-h-and-inttypes-h

File: lib/CMakeLists.txt (hunk 2 above)
Comment (carlesfernandez) [reply lib/CMakeLists.txt:153]: `#include <volk/volk_common.h>` is not used if `MSVC`, so I didn't include the flag.

**Comment by carlesfernandez** [comment]
Please check this CI in Github Actions without this change (https://github.com/carlesfernandez/volk/commit/c9cd2090ebe0b59f545c1ee20e9defd97bbe5e83/checks?check_suite_id=399828671) and with this change applied (https://github.com/carlesfernandez/volk/commit/2459583524cd39d65abb82ba2ef5c749ac854692/checks?check_suite_id=399839138)

The action is in the configure step.

File: lib/CMakeLists.txt (hunk 3 above)
Comment (balister) [reply lib/CMakeLists.txt:262]: This all looks reasonable to me. I really need to setup something in my Jenkins to run volk builds on a few toolchains

File: include/volk/volk_common.h (hunk 1 above)
Comment (jdemel) [reply include/volk/volk_common.h:90]: Since we don't need `inttypes.h` I guess, we can go for `stdint.h`. I wonder if this changed recently? I can only imagine someone used `inttypes` because of debugging.
I don't see the difference in your CI though.
Also, we should probably also set up Github Actions as an additional CI.

File: lib/CMakeLists.txt (hunk 3 above)
Comment (jdemel) [reply lib/CMakeLists.txt:262]: I thought we run ARM CI as well. Though it might be `aarch64` only. Need to check.

File: include/volk/volk_common.h (hunk 1 above)
Comment (carlesfernandez) [reply include/volk/volk_common.h:90]: The difference in the CI:

Without the change (branch `actions`), in configure step of `volk_macos` job:

```
...
-- x86* CPU detected
-- Compiler or linker missing xgetbv instruction, Overruled arch avx
-- Performing Test neon_compile_result
-- Performing Test neon_compile_result - Failed
...
```

With the change (branch`fix-avx-neon-detection-ci`), in configure step of `volk-macos` job:
```
...
-- x86* CPU detected
-- Performing Test neon_compile_result
-- Performing Test neon_compile_result - Failed
...
```

The same happens in the `volk-ubuntu` job (not related to this change in the include, though)

**Comment by carlesfernandez** [comment]
Related to ARM CI in your travis builds in arm64:

In https://travis-ci.org/gnuradio/volk/jobs/636915880?utm_medium=notification&utm_source=github_status (current upstream `master` branch):

```
...
-- Performing Test neon_compile_result
-- Performing Test neon_compile_result - Success
-- Performing Test have_neonv7_result
-- Performing Test have_neonv7_result - Failed
-- Performing Test have_neonv8_result
-- Performing Test have_neonv8_result - Failed
-- Compiler doesn't support neonv7, Overruled arch neonv7
-- Compiler doesn't support neonv8, Overruled arch neonv8
-- Available architectures: generic;softfp;hardfp;neon;64;abm;popcount;fma;orc;norc;avx2
...
```


In https://travis-ci.org/gnuradio/volk/jobs/636871816?utm_medium=notification&utm_source=github_status (CI of this PR):

```
...
-- Performing Test neon_compile_result
-- Performing Test neon_compile_result - Success
-- Performing Test have_neonv7_result
-- Performing Test have_neonv7_result - Failed
-- Performing Test have_neonv8_result
-- Performing Test have_neonv8_result - Success
-- Compiler doesn't support neonv7, Overruled arch neonv7
-- Available architectures: generic;softfp;hardfp;neon;neonv8;64;abm;popcount;fma;orc;norc;avx2
...
```

**Review by jdemel — APPROVED** [review]
This change looks good to me.
I can't verify it locally so I have to rely on CI here.

**Review by michaelld — APPROVED** [review]
LGTM

**Comment by michaelld** [comment]
CI failures have nothing to do with this PR. 2 approvals. Important fixes. Merging.
